modules/index/src/main/java/org/apache/ignite/internal/index/IndexCollector.java

```diff
@@ -0,0 +1,296 @@
+     * catalog versions.
+     *
+     * <p>Let's look at an example, let's say we have the following versions of a catalog with indexes:</p>
+     * <pre>
+     *     0: I0(A) I1(A)
+     *     1: IO(A)
+     *     2: I0(A) I2(R) I3(A)
+     *     3: I0(A)
+     *     4: I0(A)
+     * </pre>
+     *
+     * <p>Then the map will have the following values:</p>
+     * <pre>
+     *     (0, 1) -> [I1(A)]
+     *     (0, 3) -> [I1(A), I3(A)]
+     * </pre>
+     *
+     * <p>Then, when {@link #getDroppedAvailableIndexes(int, int) getting dropped available indexes}, we will return the following:</p>
+     * <pre>
+     *     (0, 0) -> []
+     *     (1, 0) -> [I1(A)]
+     *     (2, 0) -> [I1(A)]
+     *     (3, 0) -> [I1(A), I3(A)]
+     *     (4, 0) -> [I1(A), I3(A)]
+     * </pre>
+     *
+     * <p>Updated on {@link #recover() node recovery} and a catalog events processing.</p>
+     */
+    private final ConcurrentSkipListMap<TableIdCatalogVersion, List<CatalogIndexDescriptor>> droppedAvailableTableIndexes
+            = new ConcurrentSkipListMap<>();
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    /** Constructor. */
+    IndexCollector(CatalogService catalogService) {
+        this.catalogService = catalogService;
+
+        addListeners();
+    }
+
+    /** Recovers internal structures on node recovery. */
+    void recover() {
+        inBusyLock(busyLock, () -> {
+            // It is expected that the methods will be called only on recovery, when the deploy of metastore watches has not yet occurred.
+            int earliestCatalogVersion = catalogService.earliestCatalogVersion();
+            int latestCatalogVersion = catalogService.latestCatalogVersion();
+
+            // At the moment, we will only use tables from the latest version, since so far only replicas for them are started on the node.
+            for (CatalogTableDescriptor table : catalogService.tables(latestCatalogVersion)) {
+                for (int catalogVersion = earliestCatalogVersion; catalogVersion < latestCatalogVersion; catalogVersion++) {
+                    int tableId = table.id();
+                    int nextCatalogVersion = catalogVersion + 1;
+
+                    List<CatalogIndexDescriptor> tableIndexes = catalogService.indexes(catalogVersion, tableId);
+
+                    if (tableIndexes.isEmpty()) {
+                        // Table does not exist yet.
+                        continue;
+                    }
+
+                    List<CatalogIndexDescriptor> nextCatalogVersionTableIndexes = catalogService.indexes(nextCatalogVersion, tableId);
+
+                    assert !nextCatalogVersionTableIndexes.isEmpty()
+                            : String.format("Table should not be dropped: [catalogVersion=%s, tableId=%s]", nextCatalogVersion, tableId);
```

Comment:
Why should not it be dropped here? Is it because `catalogService.tables()` only returns non-dropped tables?

Reply:
Yes, improved the documentation above a bit.